SPEC.md

```diff
@@ -0,0 +1,334 @@
+
+A Nexus **caller** calls a **handler**. The handler may respond inline or return a reference for a future, asynchronous
+operation. The caller can cancel an asynchronous operation, check for its outcome, or fetch its current state. The caller
+can also specify a callback URL, which the handler uses to asynchronously deliver the result of an operation when it
+is ready.
+
+## Operation Addressability
+
+An operation is addressed using three components:
+
+- [Service Name](#service-name)
+- [Operation Name](#operation-name)
+- [Operation ID](#operation-id)
+
+> **Note:** More details regarding IDs will be provided upon finalizing the ID proposal.
+
+### Service Name
+
+Service names:
+
+- MUST not be empty.
+- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].
+
+Valid characters include:
+
+```
+ALPHA / DIGIT / "-" / "." / "_" / "~"
+```
+
+### Operation Name
+
+Operation names:
+
+- MUST not be empty.
+- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].
+
+Valid characters include:
+
+```
+ALPHA / DIGIT / "-" / "." / "_" / "~"
+```
+
+### Operation ID
+
+Operation IDs:
+
+- MUST not be empty.
+- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].
+
+Valid characters include:
+
+```
+ALPHA / DIGIT / "-" / "." / "_" / "~"
+```
+
+## Schema Definitions
+
+### Failure
+
+The `Failure` object represents outcomes that are `failed` or `canceled`. The object MUST adhere to the following JSON
+schema:
+
+```yaml
+type: object
+properties:
+  message:
+    type: string
+  details:
+    type: object
+    properties:
+      metadata:
+        type: object
+        additionalProperties:
+          type: string
+        description: |
+          String to string mapping, may contain information for decoding the data field.
+      data:
+        type: string
+        description: |
+          Arbitrary string-encoded data; use encodings like base64 for binary data transmission.
+```
+
+### OperationStarted
+
+An `OperationStarted` object indicates the commencement of an asynchronous operation. This object MUST follow the given
+JSON schema:
+
+```yaml
+type: object
+properties:
+  operationId:
+    type: string
+    description: |
+      An identifier for referencing this operation. If provided in the start request, it is echoed back.
+  callbackUrlSupported:
+    type: boolean
+    description: |
+      If the request specified a callback_url and the handler supports callbacks for this operation, this flag will be
+      set.
+      It is up to the caller to decided how to get the outcome of this call in case the handler does not support
+      callbacks.
+```
+
+### OperationInfo
+
+The `OperationInfo` object, retrieved from the get operation info API, MUST adhere to the given schema:
+
+```yaml
+type: object
+properties:
+  state:
+    enum:
+      - succeeded
+      - failed
```

Comment:
Does this include both application-level (e.g. operation involves an RPC but all retries failed) and framework-level failures that the framework caught? If so I think it might be helpful for the spec to say so explicitly.

Reply:
This is operation state, there is no difference between application and framework level. An operation is failed or it is not. How a handler represents/relays operation failure is beyond the scope of the API spec.